### PR TITLE
Tensor parallelism

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?

This PR adds support for Tensor Parallelism (including Sequence Parallelism). 
Additionally, this PR adds a debugging toolkit to track the input and output tensors during a forward pass, gradients during the backward pass and weight tensors.
Tensors can bei either normal Tensors or DTensors.  

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue in isolation
- [x] I have merged the latest version of the target branch into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [x] I have checked that all tests run through (`python tests/tests.py`)
- [x] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

**Review by flxst — CHANGES_REQUESTED** [review]
Great work! 

I think there is a problem regarding TP + GELU, see my comment in `src/modalities/models/model_factory.py`.

File: tests/fsdp2_parallelization/test_tensor_parallelism.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        with MultiProcessingCudaEnv(
+            process_group_backend=ProcessGroupBackendType.nccl,
+            global_rank=process_id,
+            local_rank=process_id,
+            world_size=world_size,
+            rdvz_port=22490,
+        ):
+            torch.manual_seed(42)
+            fsdp2_model, fsdp2_device_mesh = self._get_components(fsdp_config_path)
+            torch.manual_seed(42)
```
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/fsdp2_parallelization/test_tensor_parallelism.py:64]: If we set the manual seed again, we make sure that both models (fsdp2_model and tp_model) have the same weights.

File: src/modalities/dataloader/sampler_factory.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    device_mesh: PydanticDeviceMeshIFType
+    data_parallel_key: ParallelismDegrees
+    epoch: Annotated[int, Field(strict=True, ge=0)] = 0
+    shuffle: Optional[bool] = False
+    seed: Optional[int] = 0
+    drop_last: Literal[True] = True
+    skip_num_global_samples: Annotated[int, Field(strict=True, ge=0)] = 0
+
+
+class SamplerFactory:
```
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/dataloader/sampler_factory.py:23]: It's a component that we can instatiate: https://github.com/Modalities/modalities/blob/fb585aa9639faef93b9e437ae5a75c5eb86da19f/src/modalities/registry/components.py#L235

We need it whenever we use TP. Otherwise each rank, sees different data although we want every node to process the same data.

File: src/modalities/models/gpt2/gpt2_model.py [hunk 3]
```diff
@@ -749,8 +751,10 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         Returns:
             torch.Tensor: Output tensor.
         """
-        x = x + self.attn(self.attention_norm(x))
-        x = x + self.mlp(self.ffn_norm(x))
+        x_attn_normed = self.attention_norm(x)
+        x = x + self.attn(x_attn_normed)
+        x_ffn_normed = self.ffn_norm(x)
+        x = x + self.mlp(x_ffn_normed)
```
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/models/gpt2/gpt2_model.py:757]: I changed it for debugging purposes, so that I can place a breakpoint. I will change it back to the previous version.

File: tests/fsdp2_parallelization/test_tensor_parallelism.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from torch.distributed.tensor import DTensor, Replicate
+
+from modalities.__main__ import Main
+from modalities.config.config import ProcessGroupBackendType
+from modalities.config.pydantic_if_types import PydanticDeviceMeshIFType, PydanticFSDP2ModuleType
+from tests.end2end_tests.custom_components import MultiProcessingCudaEnv
+
+
+@pytest.mark.skipif(
+    torch.cuda.device_count() < 4,
```
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/fsdp2_parallelization/test_tensor_parallelism.py:20]: We need also the dataparallel dimension with a DP degree of >= 2. So the smallest test case is 4 GPUs.

File: tests/fsdp2_parallelization/test_tensor_parallelism.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/fsdp2_parallelization/test_tensor_parallelism.py:64]: If you seed the random generator, it always creates the same sequence of random values. 
If you then initialize two models sequentially then the first n values will be used for model and the following n values will be used for the second model. 
Thus model 1 and model 2 have different weights. 
If you seed again then you will reuse the same random sequence, as you are resetting the random generator. 
Does that make sense?

File: tests/fsdp2_parallelization/tp_test_configs/tp_config.yaml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+device_mesh:
+  component_key: device_mesh
+  variant_key: default
+  config:
+    device_type: cuda
+    data_parallel_replicate_degree: 1
+    data_parallel_shard_degree: -1
+    tensor_parallel_degree: 2
+    world_size: 4
```
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/fsdp2_parallelization/tp_test_configs/tp_config.yaml:97]: yes, and we normally do this. Since this is a test config for the the tp test (which always requires exactly 4 GPUs) I just hardcoded the number of GPUs. Probably we anyways never have to touch this again.

File: tests/fsdp2_parallelization/test_tensor_parallelism.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    @pytest.mark.parametrize(
+        "fsdp_config_path, tp_config_path",
+        [
+            (
+                Path("tests/fsdp2_parallelization/tp_test_configs/fsdp_config.yaml"),
+                Path("tests/fsdp2_parallelization/tp_test_configs/tp_config.yaml"),
+            ),
+        ],
+    )
+    def test_tp_sharding(self, fsdp_config_path: Path, tp_config_path: Path):
```
[the remark this answers is not in the document]
Comment (le1nux) [reply tests/fsdp2_parallelization/test_tensor_parallelism.py:44]: yeah TP does not work without FSDP2 anyways, that's why I took the naming. But I saw your other comment as well. I will change, the naming to FSDP2 to be consistent.

File: src/modalities/trainer.py [hunk 7]
```diff
@@ -282,7 +280,6 @@ def train(
                         "train samples/s": ResultItem(synced_num_samples_per_second, 1),
                         "train mfu (16-bit)": ResultItem(mfu_score, 2),
                         "lr mean": ResultItem(torch.tensor(lr_scheduler.get_last_lr()).mean()),
-                        "peak memory rank 0 (MB)": ResultItem(torch.tensor(peak_memory_MB), 2),
```
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/trainer.py:285]: I would add this again, within the profiling and benchmarking PR. It was more like a testing setup.

File: src/modalities/models/model_factory.py [hunk 8]
```diff
@@ -421,3 +599,96 @@ def get_gpt2_model(
...
+            "attn.v_attn": ColwiseParallel(),
+            "attn.c_proj": RowwiseParallel(output_layouts=Shard(1)),
+            "ffn_norm": SequenceParallel(),
+            "mlp": PrepareModuleInput(
+                input_layouts=(Shard(1),),
+                desired_input_layouts=(Replicate(),),
+            ),
+            "mlp.W": ColwiseParallel(),
+            "mlp.W_2": RowwiseParallel(output_layouts=Shard(1)),
+            "mlp.V": ColwiseParallel(),
```
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/models/model_factory.py:671]: Good point. I will address this.

**Comment by le1nux** [comment]
> Great work!
> 
> I think there is a problem regarding TP + GELU, see my comment in `src/modalities/models/model_factory.py`.

I have added GELU support now for TP (including unit tests)

File: src/modalities/models/model_factory.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (le1nux) [reply src/modalities/models/model_factory.py]: GELU TP support now integrated 👍

**Review by flxst — CHANGES_REQUESTED** [review]
Changes look good! Only some minor issues left, see comments.

**Review by rrutmann — APPROVED** [review]

**Review by flxst — APPROVED** [review]
LGTM
